### PR TITLE
NEW (Extension) @W-13991634@ Add class level PMD violations

### PR DESCRIPTION
**Pull request by jag-j**
This PR does the following:
1. Injects PMD's class-level suppression syntax to silence a single rule for that class (i.e., `@SuppressWarnings` syntax).
2. Identifies the specific rule to be suppressed and adds the appropriate syntax (Example: `@SuppressWarnings('PMD.ApexDoc')`).
3. If a class-level suppression tag already exists, it changes the SuppressWarnings accordingly (Example: `@SuppressWarnings('PMD.ApexDoc', 'PMD.UnusedMethod')`), instead of adding a new tag.
4. Adds the tag for Apex and Java appropriately (double-quotes for Java).
5. Identifies the correct position to add this tag by regex parsing the file for 'class', but at the same type ignoring 'class' defined as a comment or as part of a string.

### CODE REVIEW

File: src/lib/fixer.ts [hunk 1]
```diff
@@ -131,6 +138,164 @@ export class _PmdFixGenerator extends FixGenerator {
...
+            }
+        } else {
+            // If @SuppressWarnings does not exist, insert a new one
+            const newSuppression = this.generateNewSuppressionTag(suppressionRule, this.document.languageId);
+            action.edit.insert(this.document.uri, classStartPosition, newSuppression);
+        }
+    
+        action.diagnostics = [this.diagnostic];
+        action.command = {
+            command: Constants.COMMAND_REMOVE_DIAGNOSTICS_ON_SELECTED_FILE,
```
Comment (jag-j) [inline src/lib/fixer.ts:185]: FYI, I clear the diagnostics for the whole file and this is not ideal when a file has multiple classes. Keeping track of the diagnostics that would be affected by the rule added on the specific class seemed too difficult and the AC for the story did not call for it. Since this PR is already doing too many things, I could still do it as part of a separate story.

File: src/lib/fixer.ts [hunk 2]
```diff
@@ -131,6 +138,164 @@ export class _PmdFixGenerator extends FixGenerator {
...
             arguments: [this.document.uri, this.diagnostic]
         };
+
         return action;
     }
+
+    public generateClassLevelSuppression(): vscode.CodeAction {
+        // Find the end-of-line position of the class declaration where the diagnostic is found.
+        const classStartPosition = this.findClassStartPosition(this.diagnostic, this.document);
+        // const classEndOfLinePosition = 0;
```
Comment (stephen-carter-at-sf) [inline src/lib/fixer.ts:148]: Please remove commented out code.

File: src/lib/fixer.ts [hunk 3]
```diff
@@ -131,6 +138,164 @@ export class _PmdFixGenerator extends FixGenerator {
...
+        }
+    
+        // Extract text from the start to end of the class declaration to search for existing suppressions
+        const classText = this.findLineBeforeClassStartDeclaration(classStartPosition, this.document);
+        const suppressionMatch = classText.match(this.suppressionRegex);
+    
+        if (suppressionMatch) {
+            // If @SuppressWarnings exists, check if the rule is already present
+            const existingRules = suppressionMatch[1].split(',').map(rule => rule.trim());
+            if (!existingRules.includes(suppressionRule)) {
```
Comment (stephen-carter-at-sf) [inline src/lib/fixer.ts:168]: Just wondering about case sensitivity of the rules. If they are case insensitive, then we probably should add to the map a .toLowerCase() and then add it to the suppressionRule when checking for existence.

File: src/lib/fixer.ts [hunk 4]
```diff
@@ -131,6 +138,164 @@ export class _PmdFixGenerator extends FixGenerator {
...
+        }
+        return '';
+    }
+
+    /**
+     * Finds the start position of the class in the document.
+     * Assumes that the class declaration starts with the keyword "class".
+     * @returns The position at the start of the class.
+     */
+    public findClassStartPosition(diagnostic: vscode.Diagnostic, document: vscode.TextDocument): vscode.Position {
```
Comment (stephen-carter-at-sf) [inline src/lib/fixer.ts:216]: I was confused at first by this... I didn't realize until I read the code that you always return column 0. This sounds like it is the position (line, col) exactly where you find the word "class". But in reality you are finding the starting line of the class and returning the zeroth column. I recommend you update your comment to make this clear. Otherwise, I was scratching my head earlier in the code wondering how you could put a newline character immediately before the word "class".

File: src/lib/fixer.ts [hunk 5]
```diff
@@ -131,6 +138,164 @@ export class _PmdFixGenerator extends FixGenerator {
...
+            // Skip block comment in a single line
+            if (line.match(this.blockCommentEndPattern) && line.match(this.blockCommentStartPattern)) {
+                continue;
+            }
+
+            // Check if this line is the start of a block comment
+            if (line.match(this.blockCommentEndPattern)) {
+                inBlockComment = true;
+                continue;
+            }
```
Comment (stephen-carter-at-sf) [inline src/lib/fixer.ts:252]: I'm a bit concerned how strings are going to prevent these regular expressions from working as you desire them to. For example, if I have a string abc = 'some */ random string'; then it will think you just entered into a block comment (specifically because you are going backward). If you were going forward, the blockCommentStartPattern is more robust.

So I propose you first walk forward through the file and keep track of what lines are comments (since deteching comments blocks are easier going forward). Then afterwards walk backward from the diagnostic line and skip those commented lines. 

Alternatively maybe your isWithinQuotes will help here.

File: src/lib/fixer.ts [hunk 6]
```diff
@@ -131,6 +138,164 @@ export class _PmdFixGenerator extends FixGenerator {
...
+    /**
+     * Helper function to check if match is within quotes
+     * @param line 
+     * @param matchIndex 
+     * @returns 
+     */
+    public isWithinQuotes(line: string, matchIndex: number): boolean {
+        const beforeMatch = line.slice(0, matchIndex);
+        const singleQuotesBefore = (beforeMatch.match(/'/g) || []).length;
+        const doubleQuotesBefore = (beforeMatch.match(/"/g) || []).length;
```
Comment (stephen-carter-at-sf) [inline src/lib/fixer.ts:294]: Does apex allow double quotes? I thought strings must be single quoted? Not sure if you need to worry about double quotes.

File: src/lib/fixer.ts [hunk 7]
```diff
@@ -131,6 +138,164 @@ export class _PmdFixGenerator extends FixGenerator {
...
+
+        // Check if the number of quotes before the match is odd (inside quotes)
+        if (singleQuotesBefore % 2 !== 0 || doubleQuotesBefore % 2 !== 0) return true;
+
+        return false;
```
Comment (stephen-carter-at-sf) [inline src/lib/fixer.ts:299]: lower golf score:
return <conditional_statement>

File: src/lib/fixer.ts [hunk 8]
```diff
@@ -131,6 +138,164 @@ export class _PmdFixGenerator extends FixGenerator {
...
+        // Find the end-of-line position of the class declaration where the diagnostic is found.
+        const classStartPosition = this.findClassStartPosition(this.diagnostic, this.document);
+        // const classEndOfLinePosition = 0;
+
+        const action = new vscode.CodeAction(messages.fixer.supressOnClass, vscode.CodeActionKind.QuickFix);
+        action.edit = new vscode.WorkspaceEdit();
+    
+        // Determine the appropriate suppression rule based on the type of diagnostic.code
+        let suppressionRule: string;
+        if (typeof this.diagnostic.code === 'object' && this.diagnostic.code !== null && 'value' in this.diagnostic.code) {
```
Comment (jfeingold35) [inline src/lib/fixer.ts:155]: Do you need both the `=== 'object'` and `!== null` check? Seems like the first would be sufficient. Also, in JS/TS it's typical to do `== null` instead of `=== null`, because `undefined == null` and usually you want to catch both null and undefined.

File: src/lib/fixer.ts (hunk 3 above)
Comment (jfeingold35) [reply src/lib/fixer.ts:168]: Follow-up question: Is it an error case if the rule is already present, or is it valid and nothing should happen?

File: src/lib/fixer.ts (hunk 1 above)
Comment (jfeingold35) [reply src/lib/fixer.ts:185]: Clarification: Does "multiple classes" mean multiple outer classes, or an outer class with inner classes? I know Apex can't have multiple outer classes, and I'm pretty sure that the annotation will only apply to outer classes, though I haven't checked that.

File: src/lib/fixer.ts (hunk 5 above)
Comment (jfeingold35) [reply src/lib/fixer.ts:252]: I agree with Stephen. I'd add tests for this kind of corner case if you can.

File: src/lib/fixer.ts (hunk 6 above)
Comment (jfeingold35) [reply src/lib/fixer.ts:294]: Strings have to be single-quoted. I genuinely don't remember if you're allowed to use double-quotes as string boundaries.
BUT you might want to check whether the string containing escaped single-quotes throws off the algorithm. E.g., `'this is a string. it\'s got a single-quote inside it'`.

File: src/lib/fixer.ts [hunk 9]
```diff
@@ -131,6 +138,164 @@ export class _PmdFixGenerator extends FixGenerator {
...
+     * @param matchIndex 
+     * @returns 
+     */
+    public isWithinQuotes(line: string, matchIndex: number): boolean {
+        const beforeMatch = line.slice(0, matchIndex);
+        const singleQuotesBefore = (beforeMatch.match(/'/g) || []).length;
+        const doubleQuotesBefore = (beforeMatch.match(/"/g) || []).length;
+
+        // Check if the number of quotes before the match is odd (inside quotes)
+        if (singleQuotesBefore % 2 !== 0 || doubleQuotesBefore % 2 !== 0) return true;
```
Comment (jfeingold35) [inline src/lib/fixer.ts:297]: Pet peeve: I'm broadly opposed to single-line/braceless `if`s.

File: src/lib/fixer.ts (hunk 1 above)
Comment (jag-j) [reply src/lib/fixer.ts:193]: Didn't realize apex can't have multiple outer classes. I have changed the code and now this applies only to the outer class.

File: src/lib/fixer.ts (hunk 3 above)
Comment (jag-j) [reply src/lib/fixer.ts:175]: I have added case insensitivity in the regex and added tests. It is an error case when a rule is already present and violations are present. One scenario I can think of is when user ran the scanner without the rule and manually added the rule and then tried to run the suppress at the class level. I'm not handling that part.

File: src/lib/fixer.ts (hunk 5 above)
Comment (jag-j) [reply src/lib/fixer.ts]: Took care of this.

File: src/lib/fixer.ts (hunk 7 above)
Comment (jag-j) [reply src/lib/fixer.ts]: Done.

File: src/lib/fixer.ts (hunk 8 above)
Comment (jag-j) [reply src/lib/fixer.ts]: Fixed.

File: src/lib/fixer.ts (hunk 3 above)
Comment (stephen-carter-at-sf) [reply src/lib/fixer.ts:175]: Adding case insensitivity in the regex doesn't help with this line does it existingRules.includes(suppressionRule)) ?

I'd recommend something like
            const existingRules = suppressionMatch[1].split(',').map(rule => rule.trim().toLowerCase());
            if (!existingRules.includes(suppressionRule.toLowerCase())) {

**Review by stephen-carter-at-sf — APPROVED** [review]

File: src/lib/fixer.ts (hunk 3 above)
Comment (jag-j) [reply src/lib/fixer.ts:175]: As discussed offline, this has an unintended side effect. It changes all existing rules to lower case since we concatenate the existing rules (which were made to be lower case for comparison) to new rules. So, we decided to leave this as is.
